### DISCUSSION TITLE
How to configure the hostRules for AWS ECR on the renovate-on-prem

### DISCUSSION THREAD

**Discussion by tata9001** [renovatebot/renovate#10062] [Request Help]
How can I configure the hostRule for the ECR when the renovate service is running on the AWS ECS Fargate with the ECR related permission?
```js
    "hostRules": [
    {
      "hostType": "docker",
      "matchHost": "https://xxx.dkr.ecr.ap-southeast-2.amazonaws.com",
    },
    {
      "hostType": "docker",
      "matchHost": "https://index.docker.io",
      "username": "xxx",
      "password": process.env.DOCKER_HUB_PWD,
    },
  ],

```
I have tried this way, but it doesn't work.

From https://github.com/renovatebot/renovate/issues/9766

**Comment by rarkins** [comment]
Your first rule doesn't do anything. What were you trying to do?

**Reply by tata9001** [reply]
My renovate service is running on the AWS ECS Fargate with the ECR permission.
```yaml
    policy_statements:
      ecr:
        effect: Allow
        action:
          - ecr:BatchCheckLayerAvailability
          - ecr:BatchGetImage
          - ecr:GetAuthorizationToken
          - ecr:GetDownloadUrlForLayer
        resource: "*"
```
So I don't need to provide the credentials to [ECR](https://github.com/fgreinacher/renovate/blob/22cfed8cf20070d1680d367cdc4b60eea19cc7bd/lib/datasource/docker/index.ts#L117-L122), the AWS SDK should fetch it from metadata server automatically.

I add this config is just letting renovate know you have the power to fetch ECR.

---

Actually I have tested to add an IAM user here, it also doesn't work on the [renovate-on-prem](https://github.com/whitesource/renovate-on-prem)

```js
 "hostRules": [
    {
      "hostType": "docker",
      "matchHost": "https://xxx.dkr.ecr.ap-southeast-2.amazonaws.com",
      "username": "{iam-user-accessKeyId}",
      "password": "{iam-user-secretAccessKey}",
    },
    {
      "hostType": "docker",
      "matchHost": "https://index.docker.io",
      "username": "xxx",
      "password": process.env.DOCKER_HUB_PWD,
    },
  ],
```

**Reply by rarkins** [reply]
> I add this config is just letting renovate know you have the power to fetch ECR.

That's not how hostRules work.

**Reply by tata9001** [reply]
So which way should I use it? Can you give me an example? Thanks very much. 🙏

**Reply by rarkins** [reply]
Right now I can only direct you to the relevant source code, which starts here: https://github.com/renovatebot/renovate/blob/52e694bec9f2100fdd921903af0ef76ff5f44183/lib/datasource/docker/common.ts#L18

You can see that if the regex is matched then a function is called to fetch ECR credentials and add them to the request.

If username/password credentials are supplied via a hostRule then they'd be passed to the AWS ECR lib, but otherwise the lib should find them from the system if it can reach the internal AWS endpoint.

**Reply by fgreinacher** [reply]
I’m not exactly sure that the AWS SDK automatically connects to the metadata service. The documentation at https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/modules/_aws_sdk_credential_provider_imds.html indicates that this needs to be configured explicitly.

**Reply by tata9001** [reply]
@fgreinacher Based on [this document](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html), I think it will.


Due to your [ECR fix](https://github.com/renovatebot/renovate/pull/9767) is still not baked into [renovate-on-prem](https://github.com/whitesource/renovate-on-prem/issues/149), I use some tricky to fix it by dockerfile:
```
FROM whitesource/renovate:1.3.1

USER root

COPY config.js /usr/src/app/config.js

RUN yarn add renovate@25.4.7

USER 1000
```

And with this config,
```js
  hostRules: [
    {
      hostType: "docker",
      matchHost: "https://xxx.dkr.ecr.ap-southeast-2.amazonaws.com",
      username: "xxx",
      password: "xxx",
    },
    {
      hostType: "docker",
      matchHost: "https://index.docker.io",
      username: "aaa",
      password: "xxxx"
    },
  ],
```


But it still cannot work, can you provide your ECR related config about this?

(I have read few times about the code that @rarkins pointed, still cannot figure out why it cannot work.)

**Reply by fasterthanlime** [reply]
Did you mean to tag @fgreinacher ?

**Reply by fgreinacher** [reply]
Your document refers to the v2 AWS SDK, but Renovate is using v3.

I don’t have our config at hand (vacation) but I think we’re passing the credentials of a standard IAM user via username/password.

**Reply by tata9001** [reply]
```
  hostRules: [
    {
      hostType: "docker",
      matchHost: "https://xxx.dkr.ecr.ap-southeast-2.amazonaws.com",
      username: "xxx",
      password: "xxx",
    },
    {
      hostType: "docker",
      matchHost: "https://index.docker.io",
      username: "aaa",
      password: "xxxx"
    },
  ],
```

In the latest [renovate-on-prem](https://github.com/whitesource/renovate-on-prem/releases/tag/1.4.2), this config worked.

**Comment by BastianVoigt** [comment]
This doesn't seem to work with temporary credentials, where the `AWS_SESSION_TOKEN` would be needed as a third parameter, right? I tried to get it to work with the EC2 instance role. Is it possible to combine it somehow with the CLI command `aws ecr get-login-password` ?

I have tried it like this, but no success. I passed this environment variable to the renovate process:

    DOCKER_PASSWORD=$(aws ecr get-login-password --region eu-central-1)

And then referred to it in the config:

     {
      hostType: "docker",
      matchHost: "https://12345678910.dkr.ecr.eu-central-1.amazonaws.com",
      username: "AWS",
      password: process.env.DOCKER_PASSWORD
    }

What am I doing wrong?

**Reply by BastianVoigt** [reply]
By the way, is the "https://" prefix correct or should I remove that?

**Reply by viceice** [reply]
prefix doesn't matter.

you can pass the temporary key as token to the same hostrule.

https://github.com/renovatebot/renovate/blob/e635f0e2961a2cd4ff6d095735c120da802be37d/lib/modules/datasource/docker/ecr.ts#L21

**Reply by BastianVoigt** [reply]
DEBUG: GET https://12345678910.dkr.ecr.eu-central-1.amazonaws.com/v2/my-image/tags/list?n=1000 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=51) (repository=GLOBIT/my-image)

**Reply by sgorshkov85** [reply]
@BastianVoigt  @viceice @rarkins   Hello. Have you found a way to use a temp AWS ECR token? I tried several ways to set hostrules. 
 Such as:
```
       {
        hostType: 'docker',
        matchHost: '11111111.dkr.ecr.eu-central-1.amazonaws.com',
        username: 'AWS',
        token: process.env.ECR_AUTH_TOKEN
      },
```

**Reply by viceice** [reply]
that's currently not supported, PR welcome.

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/blob/58cc2c073738daca066ed87130099f59630c6d9f/lib/modules/datasource/docker/common.ts#L99

**Reply by sgorshkov85** [reply]
Thanks for the feedback. I was able to figure out how to use the temporary credentials. It might be useful for others.
 In your CI you could get temp secrets with the next command(IAM role attached to the runner, slave must have enough permission to assume the role):
``` 
  $(printf "export AWS_ACCESS_KEY_ID=%s AWS_SECRET_ACCESS_KEY=%s AWS_SESSION_TOKEN=%s" $(aws sts assume-role --role-arn arn:aws:iam::111111111:role/my-iam-role --role-session-name MySessionName --query "Credentials.[AccessKeyId,SecretAccessKey,SessionToken]" --output text))
```

 Configuration in config.js:
``` 
     {
        hostType: 'docker',
        matchHost: '11111111.dkr.ecr.eu-central-1.amazonaws.com',
        usernname: process.env.AWS_ACCESS_KEY_ID,
        password:  process.env.AWS_SECRET_ACCESS_KEY,
        token:     process.env.AWS_SESSION_TOKEN
      },
``` 

@viceice Sure I'll try to create PR when I have enough time.

**Comment by SS-SamiAhonen** [comment]
Can some one share AWS configs relating this? I do not find detailed documentation and with IAM user defined into configs (renovate.json) user pwd decryption works, but still getting this:

DEBUG: ECR getAuthorizationToken error (repository=org/repo)
 WARN: No docker auth found - returning (repository=org/repo)

**Comment by jjlakis** [comment]
Following works for me. In config.json:
```
  "packageRules": [
    {
      "datasources": ["docker"],
      "registryUrls": ["https://registry.dkr.ecr.eu-west-3.amazonaws.com"]
    }
  ]
```
and Gitlab CI job:
```
renovate:
  stage: update
  image: renovate/renovate
  variables:
    RENOVATE_REPOSITORIES: my-gitlab-group/my-gitlab-repo
    RENOVATE_PLATFORM: gitlab
    RENOVATE_PR_HOURLY_LIMIT: 0
    LOG_LEVEL: debug
    AWS_ROLE_SESSION_NAME: $CI_PIPELINE_IID
    AWS_ROLE_ARN: role-arn
    AWS_WEB_IDENTITY_TOKEN_FILE: /home/ubuntu/token
  id_tokens:
    TOKEN:
      aud: https://audience.com
  before_script:
    - echo $TOKEN > $AWS_WEB_IDENTITY_TOKEN_FILE
  script:
    - renovate
```
(role is configured to be assumed with WebIdentity from GL)

**Reply by ronberna** [reply]
Does renovate work with IRSA (IAM Roles for Service Accounts)? I'm having issues trying to connect to ECR.  I have my hostRules set to:
```
"hostRules": [
    {
        "hostType": "docker",
        "matchHost": "accountid.dkr.ecr.us-east-1.amazonaws.com"
    }
 ] 
```


but I don't see any entries in the logs that is trying to authenticate to ECR or get a token.  I validated that the regex in the code


`export const ecrRegex = /\d+\.dkr\.ecr\.([-a-z0-9]+)\.amazonaws\.com/; `


 matches my ecr repository url, but the code to get the ECR auth token is not getting executed.  Can anyone point to what could be causing this?
